runtime/lua/vim/lsp/semantic_tokens.lua

```diff
@@ -0,0 +1,233 @@
+      local hl = base_map[key]
+      if hl then
+	local prefixed
+	if type(hl) == "table" then
+	  prefixed = {}
+	  for k, v in pairs(hl) do
+	    local prefixed_v = ft..v
+	    vim.cmd("highlight default link " .. prefixed_v .. " " .. v)
+	    prefixed[k] = prefixed_v
+	  end
+	else
+	  prefixed = ft..hl
+	  vim.cmd("highlight default link " .. prefixed .. " " .. hl)
+	end
+	rawset(inner_table, key, prefixed)
+	return prefixed
+      end
+      return hl
+    end})
+    rawset(outer_table, ft, rtn)
+    return rtn
+  end})
+end
+
+local token_cache = make_cache(M.token_map)
+local modifiers_cache = make_cache(M.modifiers_map)
+
+---@private
+local function highlight(buf, token, hl)
+  vim.highlight.range(buf, ns, hl, {token.line, token.start_char}, {token.line, token.start_char + token.length})
+end
+
+---@private
+local function highlight_token(buf, token, ft_token_cache, ft_modifiers_cache)
+  local hl = ft_token_cache[token.type]
+  if hl then
+    highlight(buf, token, hl)
+  end
+  for _, m in pairs(token.modifiers) do
+    hl = ft_modifiers_cache[m]
+    -- modifiers can have a per-type mapping
+    -- e.g. readonly = { variable = "ReadOnlyVariable" }
+    if type(hl) == "table" then
+      hl = hl[token.type]
+    end
+    if hl then
+      highlight(buf, token, hl)
+    end
+  end
+end
+
+--- |lsp-handler| for the method `textDocument/semanticTokens/full`
+---
+function M.on_full(err, response, ctx, _)
+  if not ctx.bufnr or not ctx.client_id then return end
+  active_requests[ctx.bufnr] = false
+  vim.api.nvim_buf_clear_namespace(ctx.bufnr, ns, 0, -1)
+  -- if tick is has changed our response is outdated!
+  if err or last_tick[ctx.bufnr] ~= vim.api.nvim_buf_get_changedtick(ctx.bufnr) then
+    semantic_tokens[ctx.bufnr][ctx.client_id] = {}
+    return
+  end
+  local client = vim.lsp.get_client_by_id(ctx.client_id)
+  if not client then
+    return
+  end
+  local legend = client.server_capabilities.semanticTokensProvider.legend
+  local token_types = legend.tokenTypes
+  local token_modifiers = legend.tokenModifiers
+  local data = response.data
+  local ft = vim.api.nvim_buf_get_option(ctx.bufnr, "filetype")
+  local ft_token_cache = token_cache[ft]
+  local ft_modifiers_cache = modifiers_cache[ft]
+
+  local tokens = {}
+  local line, start_char = nil, 0
+  for i = 1, #data, 5 do
+    local delta_line = data[i]
+    line = line and line + delta_line or delta_line
+    local delta_start = data[i + 1]
+    start_char = delta_line == 0 and start_char + delta_start or delta_start
+
+    -- data[i+3] +1 because Lua tables are 1-indexed
+    local token_type = token_types[data[i + 3] + 1]
+    local modifiers = modifiers_from_number(data[i + 4], token_modifiers)
+
+    local token = {
+      line = line,
+      start_char = start_char,
+      length = data[i + 2],
+      type = token_type,
+      modifiers = modifiers
+    }
+    tokens[line + 1] = tokens[line + 1] or {}
```

Comment:
the look-up per line is actually not necessary. It just was is the previous implementation and it might be useful for someone searching in the tokens

Reply:
Can you explain more?

Reply:
I think I'd tend to remove the state/caching for now if we don't need the cache for a highlighting implementation. It's easy to add later once we've a concrete use-case. Making changes to the structure/interfaces is more problematic if we get it wrong.